src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -533,9 +541,15 @@ public void setProxyHandlerFactory(final ProxyHandlerFactory proxyHandlerFactory
                     new IllegalStateException("Client's event loop group has been shut down and cannot be restarted."));
         } else {
             synchronized (this.bootstrap) {
+                final ApnsClientMetricsListener metricsListener = this.metricsListener;
+
                 // We only want to begin a connection attempt if one is not already in progress or complete; if we already
                 // have a connection future, just return the existing promise.
                 if (this.connectionReadyPromise == null) {
+                    if (metricsListener != null) {
```

Comment:
I would be inclined to create a NullMetricsListener class which just contains no-ops, and use that by default, in order to avoid the need to null-check this on each call (as in NullMetrics in https://github.com/relayrides/pushy/pull/222/files#diff-bd45fb4e15d6ed613f0e6f1e1d5a193fR6 ).